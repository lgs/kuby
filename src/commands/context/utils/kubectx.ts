import { EOL } from 'os';

import { exec } from '../../../utils/exec';

export function getCurrentContext(): Promise<string> {
  return exec(`kubectl config view -o=jsonpath='{.current-context}'`);
}

export async function getContexts(): Promise<string[]> {
  const value = await exec('kubectl config get-contexts -o=name');
  return value
    .split(EOL)
    .filter(ctx => !!ctx)
    .sort();
}
