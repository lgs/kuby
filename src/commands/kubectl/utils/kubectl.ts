import { ensureDir, lstatSync, pathExists, readdir, readJson, writeJson } from 'fs-extra';
import { get } from 'got';
import { homedir, platform } from 'os';
import { join } from 'path';
import { clean, rcompare } from 'semver';

const spinner = require('ora')();

const initialUrl =
  'https://api.github.com/repos/kubernetes/kubernetes/releases?per_page=100';

export const kubectlInstallDir = join(
  homedir(),
  '.kube',
  'k8s-helpers',
  'kubectl',
);

export const kubectlVersionsFile = join(
  homedir(),
  '.kube',
  'k8s-helpers',
  'kubectl',
  'versions',
);

export const kubectlDownloadUrl = (
  version: string,
  os: 'linux' | 'darwin' | 'windows',
) =>
  `https://storage.googleapis.com/kubernetes-release/release/v${version}/bin/${os}/amd64/kubectl${
    os === 'windows' ? '.exe' : ''
  }`;

export function getOs(): 'linux' | 'darwin' | 'windows' {
  switch (platform()) {
    case 'darwin':
      return 'darwin';
    case 'win32':
      return 'windows';
    default:
      return 'linux';
  }
}

export async function getLocalVersions(): Promise<string[]> {
  await ensureDir(kubectlInstallDir);
  return ((await readdir(kubectlInstallDir))
    .filter(path => lstatSync(join(kubectlInstallDir, path)).isDirectory())
    .map(v => clean(v))
    .filter(Boolean) as string[]).sort(rcompare);
}

export async function downloadRemoteVersions(): Promise<string[]> {
  const versions: string[] = [];

  let getUrl: string | null = initialUrl;
  spinner.start(`Downloading from: ${getUrl}`);
  while (getUrl) {
    spinner.text = `Downloading from: ${getUrl}`;
    const result = await get(getUrl);

    versions.push(
      ...JSON.parse(result.body).map((release: any) => release.tag_name),
    );

    const { link } = result.headers;
    const match = /<(.*?)>; rel="next"/g.exec((link as any) || '');

    getUrl = match ? match[1] : null;
  }
  spinner.succeed();

  return versions.map(v => clean(v)).filter(Boolean) as string[];
}

export async function getRemoteVersions(): Promise<string[]> {
  if (await pathExists(kubectlVersionsFile)) {
    return await readJson(kubectlVersionsFile, { encoding: 'utf8' });
  }

  const versions = await downloadRemoteVersions();
  await writeJson(kubectlVersionsFile, versions, { encoding: 'utf8' });

  return await readJson(kubectlVersionsFile, { encoding: 'utf8' });
}
