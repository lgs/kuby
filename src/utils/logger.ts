import chalk from 'chalk';

export enum LogLevel {
  debug,
  info,
  warn,
  error,
  output,
}

export class Logger {
  public static level: LogLevel = LogLevel.info;

  constructor(private prefix: string = '') {}

  public log(level: LogLevel, message: string): void {
    if (level < Logger.level) {
      return;
    }
    this.getLogFunction(level)(`${this.getLevelPrefix(level)}${message}`);
  }

  public debug(message: string): void {
    this.log(LogLevel.debug, message);
  }

  public info(message: string): void {
    this.log(LogLevel.info, message);
  }

  public output(message: string): void {
    this.log(LogLevel.output, message);
  }

  public success(message: string): void {
    this.log(LogLevel.info, chalk.green(message));
  }

  public warn(message: string): void {
    this.log(LogLevel.warn, chalk.yellow(message));
  }

  public error(message: string): void {
    this.log(LogLevel.error, chalk.red(message));
  }

  private getLevelPrefix(level: LogLevel): string {
    if (!this.prefix) {
      return '';
    }
    switch (level) {
      case LogLevel.debug:
        return chalk.grey(`[${this.prefix}]: `);
      case LogLevel.warn:
        return chalk.yellow(`[${this.prefix}]: `);
      case LogLevel.error:
        return chalk.red(`[${this.prefix}]: `);
      default:
        return chalk.blue(`[${this.prefix}]: `);
    }
  }

  private getLogFunction(level: LogLevel): (message: string) => void {
    switch (level) {
      case LogLevel.debug:
        return console.debug;
      case LogLevel.warn:
        return console.warn;
      case LogLevel.error:
        return console.error;
      default:
        return console.info;
    }
  }
}
