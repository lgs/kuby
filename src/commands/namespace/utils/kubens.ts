import { exec } from '../../../utils/exec';
import { getCurrentContext } from '../../context/utils/kubectx';

export async function getCurrentNamespace(): Promise<string> {
  const currentContext = await getCurrentContext();
  return await exec(
    `kubectl config view -o=jsonpath='{.contexts[?(@.name=="${currentContext}")].context.namespace}'`,
  );
}

export async function getNamespaces(): Promise<string[]> {
  const value = await exec(
    `kubectl get namespaces -o=jsonpath='{range .items[*].metadata.name}{@}{"\\n"}{end}'`,
  );
  return value
    .split('\n')
    .filter(ns => !!ns)
    .sort();
}

export async function getServiceAccountsForNamespace(
  namespace: string,
): Promise<string[]> {
  const value = await exec(
    `kubectl get serviceaccounts -n ${namespace} -o=jsonpath='{range .items[*].metadata.name}{@}{"\\n"}{end}'`,
  );
  return value
    .split('\n')
    .filter(sa => !!sa)
    .sort();
}
