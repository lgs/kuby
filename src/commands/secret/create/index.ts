import { V1ObjectMeta, V1Secret } from '@kubernetes/client-node';
import { Arguments, Argv, CommandModule } from 'yargs';

import { RootArguments } from '../../../root-arguments';
import { ExitCode } from '../../../utils/exit-code';
import { KubernetesApi } from '../../../utils/kubernetes-api';
import { Logger } from '../../../utils/logger';

type SecretCreateArguments = RootArguments & {
  name: string;

  /**
   * Array of already splitted and converted data for the secret.
   */
  data: { name: string; value: string }[];
};

export const secretCreateCommand: CommandModule<RootArguments, SecretCreateArguments> = {
  command: 'create <name> <data...>',
  describe: 'Create a generic secret.',

  builder: (argv: Argv<RootArguments>) =>
    (argv
      .example(
        'kuby secret create my-secret foo=bar',
        'This creates an secret with the name my-secret and the data foo: bar. (bar will be base64 encoded)',
      )
      .example(
        'kuby secret create my-secret foo=bar this=that good="morning sir"',
        'This creates an secret with the name my-secret and the given data (encoded).',
      )
      .positional('name', {
        description: 'Name of the secret to create',
        type: 'string',
      })
      .positional('data', {
        description: 'Array of secret data in the form of: key=value key=value key=value',
        coerce: (data: string[]) =>
          data
            .map(str => str.split('='))
            .map(([name, ...values]) => ({
              name,
              value: values.join('='),
            }))
            .filter(obj => obj.name && obj.value),
        type: 'string',
      }) as unknown) as Argv<SecretCreateArguments>,

  async handler(args: Arguments<SecretCreateArguments>): Promise<void> {
    const logger = new Logger('secrets');
    logger.debug('Create secret with data.');

    if (args.data.length <= 0) {
      logger.error('No data given. Must be in format key=value or key="value value". Aborting.');
      process.exit(ExitCode.error);
      return;
    }

    const api = KubernetesApi.fromDefault();

    logger.info(`Create secret with name "${args.name}".`);
    logger.debug(`Add ${args.data.length} data keys.`);
    try {
      logger.startSpinner('Creating...');
      await api.core.createNamespacedSecret(api.currentNamespace, {
        ...new V1Secret(),
        apiVersion: 'v1',
        kind: 'Secret',
        type: 'Opaque',
        metadata: {
          ...new V1ObjectMeta(),
          name: args.name,
        },
        stringData: args.data.reduce(
          (obj, cur) => {
            obj[cur.name] = cur.value;
            return obj;
          },
          {} as { [key: string]: string },
        ),
      });
      logger.stopSpinner();
      logger.success('Secret created.');
    } catch ({ body: { message } }) {
      logger.stopSpinner();
      logger.error('There was an error during the creation of the secret:');
      logger.error(message);
      process.exit(ExitCode.error);
      return;
    }
  },
};
