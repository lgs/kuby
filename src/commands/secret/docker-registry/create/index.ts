import { async } from 'fast-glob';
import { pathExists } from 'fs-extra';
import { prompt, Separator } from 'inquirer';
import { EOL } from 'os';
import { join, parse } from 'path';
import { Arguments, Argv, CommandModule } from 'yargs';

import { RootArguments } from '../../../../root-arguments';
import { Crypto } from '../../../../utils/crypto';
import { exec } from '../../../../utils/exec';
import { ExitCode } from '../../../../utils/exit-code';
import { Filepathes } from '../../../../utils/filepathes';
import { Logger } from '../../../../utils/logger';
import { RcFile } from '../../../../utils/rc-file';
import { simpleConfirm } from '../../../../utils/simple-confirm';
import { spawn } from '../../../../utils/spawn';

interface PromptAnswers {
  server: string;
  user: string;
  email: string;
  password: string;
  saveSecret?: boolean;
  saveSecretName?: string;
}

interface SecretDockerRegistryCreateArguments extends Arguments, RootArguments {
  name: string;
  noInteraction: boolean;
  from?: string;
  server?: string;
  user?: string;
  email?: string;
  password?: string;
}

async function loadDockerSecret(
  args: SecretDockerRegistryCreateArguments,
  path: string,
): Promise<void> {
  const secret = await Crypto.load<PromptAnswers>(path);
  args.server = secret.server;
  args.user = secret.user;
  args.email = secret.email;
  args.password = secret.password;
}

export const secretDockerRegistryCreateCommand: CommandModule = {
  command: 'create <name>',
  describe:
    `Create a docker registry secret with the name "<name>".${EOL}` +
    'For each option that is omitted, the user will be asked to enter the information.',

  builder: (argv: Argv) =>
    argv
      .example(
        'k8s secret docker-registry create test --server hub.docker.com --user myaccount --email foo@bar.ch --password asdf',
        'This command creates a dockerconfigjson secret for the server hub.docker.com with the given credentials.',
      )
      .example(
        'k8s secret docker-registry create test',
        'This command creates a dockerconfigjson secret named test, and the user will be asked for the other information.',
      )
      .positional('name', {
        description: 'The name of the secret.',
      })
      .option('n', {
        alias: 'no-interaction',
        boolean: true,
        default: false,
        description: 'No interaction mode, use options or throw error.',
      })
      .option('from', {
        string: true,
        description: 'Use an already saved dockersecret as template',
      })
      .option('server', {
        string: true,
        description: 'The docker registry server (e.g. hub.docker.com).',
      })
      .option('user', {
        string: true,
        description: 'Username of the docker user.',
      })
      .option('email', {
        string: true,
        description: 'Email for the given user.',
      })
      .option('password', {
        string: true,
        description: 'Password for the given user.',
      }),

  async handler(args: SecretDockerRegistryCreateArguments): Promise<void> {
    const logger = new Logger('secrets');
    logger.debug('Create docker secret.');

    const secrets = (await exec(
      `kubectl ${RcFile.getKubectlArguments(args, [
        'get',
        'secrets',
        '-o',
        `jsonpath='{range .items[*]}{@.metadata.name}{"\\n"}{end}'`,
      ]).join(' ')}`,
    )).split('\n');

    if (secrets.includes(args.name)) {
      logger.error(`The secret "${args.name}" already exists.`);
      if (args.ci) {
        logger.warn('CI Mode: existing secret does not fail command.');
        return;
      }
      process.exit(ExitCode.error);
      return;
    }

    if (args.from) {
      const path = join(Filepathes.dockerSecretPath, args.from);
      if (await pathExists(path)) {
        await loadDockerSecret(args, path);
      } else {
        logger.error(`The file ${path} does not exist. Cannot use template.`);
      }
    } else if (!!!args.from && !args.noInteraction) {
      const secretFiles = await async<string>(['./*'], {
        cwd: Filepathes.dockerSecretPath,
      });
      const selected = (await prompt([
        {
          type: 'list',
          name: 'file',
          message:
            'Use an existing secret as value template, or create a new one?',
          choices: [
            { value: 'new', name: 'create new' },
            new Separator(),
            ...secretFiles.map(file => ({
              value: file,
              name: parse(file).name,
            })),
          ],
        },
      ])) as { file: string };
      if (selected.file !== 'new') {
        const path = join(Filepathes.dockerSecretPath, selected.file);
        args.from = selected.file;
        await loadDockerSecret(args, path);
      }
    }

    if (
      args.noInteraction &&
      (!args.server || !args.user || !args.email || !args.password)
    ) {
      logger.error(
        'No interaction mode used but not all information present (server, user, mail, password).',
      );
      process.exit(ExitCode.error);
      return;
    }

    const questions = [
      {
        type: 'input',
        name: 'server',
        message: `The server url of the registry?`,
        when: () => !args.noInteraction && !!!args.server,
        validate: (input: string) => !!input || 'Please enter a server url.',
      },
      {
        type: 'input',
        name: 'user',
        message: `The user of the registry?`,
        when: () => !args.noInteraction && !!!args.user,
        validate: (input: string) => !!input || 'Please enter a username.',
      },
      {
        type: 'input',
        name: 'email',
        message: `The email of the registry user?`,
        when: () => !args.noInteraction && !!!args.email,
        validate: (input: string) => !!input || 'Please enter an email address.',
      },
      {
        type: 'password',
        name: 'password',
        message: `The password of the registry user?`,
        when: () => !args.noInteraction && !!!args.password,
        validate: (input: string) => !!input || 'Please enter a password.',
      },
      {
        type: 'confirm',
        name: 'saveSecret',
        message:
          'Save the given secret as template (encrypted, to be used with --from in the future)?',
        default: false,
        when: () => !args.noInteraction && !!!args.from,
      },
      {
        type: 'input',
        name: 'saveSecretName',
        message: 'Name of the file?',
        default: args.name,
        when: (answers: PromptAnswers) =>
          !args.noInteraction && !!!args.from && !!answers.saveSecret,
        validate: (input: string) => !!input || 'Please enter a filename.',
      },
    ];

    const createArgs = {
      ...args,
      ...((await prompt(questions)) as PromptAnswers),
    };

    const code = await spawn(
      'kubectl',
      RcFile.getKubectlArguments(args, [
        'create',
        'secret',
        'docker-registry',
        createArgs.name,
        `--docker-server=${createArgs.server}`,
        `--docker-username=${createArgs.user}`,
        `--docker-email=${createArgs.email}`,
        `--docker-password=${createArgs.password}`,
      ]),
    );
    if (code !== 0) {
      logger.error(
        'An error happend during the kubectl create secret command.',
      );
      process.exit(ExitCode.error);
      return;
    }

    logger.success(`Docker-Registry secret "${args.name}" created.`);

    if (createArgs.saveSecret) {
      const name = createArgs.saveSecretName || createArgs.name;
      const path = join(Filepathes.dockerSecretPath, name);
      if (
        (await pathExists(path)) &&
        !(await simpleConfirm(
          `Secret with name "${name}" already exists, overwrite?`,
          false,
        ))
      ) {
        return;
      }
      await Crypto.save(path, {
        server: createArgs.server,
        user: createArgs.user,
        email: createArgs.email,
        password: createArgs.password,
      });
      logger.success(`Secret saved under ${path}.`);
    }
  },
};
