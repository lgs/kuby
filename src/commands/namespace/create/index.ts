import chalk from 'chalk';
import { ensureFile, pathExists, readFile, writeFile } from 'fs-extra';
import { prompt } from 'inquirer';
import { EOL } from 'os';
import { Arguments, Argv, CommandModule } from 'yargs';

import { RootArguments } from '../../../root-arguments';
import { datasubst } from '../../../utils/envsubst';
import { exec } from '../../../utils/exec';
import { ExitCode } from '../../../utils/exit-code';
import { Filepathes } from '../../../utils/filepathes';
import { RcFile } from '../../../utils/rc-file';
import { simpleConfirm } from '../../../utils/simple-confirm';
import { spawn } from '../../../utils/spawn';
import { getCurrentContext } from '../../context/utils/kubectx';
import { namespaceKubeConfigCommand } from '../kube-config';
import { getNamespaces } from '../utils/kubens';

interface NamespaceCreateArguments extends Arguments, RootArguments {
  name: string;
  base64: boolean;
  noInteraction: boolean;
}

interface PromptAnswers {
  createServiceAccount: boolean;
  serviceAccountName: string;
  role: string;
  saveRole: boolean;
}

const serviceAccount = (namespace: string, saName: string) => `apiVersion: v1
kind: ServiceAccount
metadata:
  name: ${saName}
  namespace: ${namespace}
`;

const rolebinding = (
  namespace: string,
  saName: string,
  roleName: string,
) => `apiVersion: rbac.authorization.k8s.io/v1
kind: RoleBinding
metadata:
  name: ${saName}
  namespace: ${namespace}
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: Role
  name: ${roleName}
subjects:
- kind: ServiceAccount
  name: ${saName}
  namespace: ${namespace}
`;

const defaultRoleYml = `# possible variables: NAME (name of the namespace)
apiVersion: rbac.authorization.k8s.io/v1beta1
kind: Role
metadata:
  name: $\{NAME\}-deploy
  namespace: $\{NAME\}
rules:
  - apiGroups:
      - ''
      - apps
      - batch
      - certmanager.k8s.io
      - extensions
    resources:
      - certificates
      - cronjobs
      - configmaps
      - deployments
      - ingresses
      - jobs
      - persistentvolumeclaims
      - pods
      - secrets
      - services
    verbs:
      - '*'
`;

export const namespaceCreateCommand: CommandModule = {
  command: 'create <name>',
  describe:
    'Create a new kubernetes namespace (with optional service account and kubeconfig).',

  builder: (argv: Argv) =>
    argv
      .positional('name', {
        description: 'Kubernetes namespace name.',
        type: 'string',
      })
      .option('b', {
        alias: 'base64',
        default: false,
        description: 'Output the kube-config encoded in base64.',
      })
      .option('n', {
        alias: 'no-interaction',
        boolean: true,
        default: false,
        description: 'No interaction mode, use default answers.',
      }),

  async handler(args: NamespaceCreateArguments): Promise<void> {
    console.group(chalk.underline('Create kubernetes namespace.'));

    const context = await getCurrentContext();
    const namespaces = await getNamespaces();

    if (namespaces.includes(args.name)) {
      console.error(chalk.red(`The namespace "${args.name}" already exists.`));
      console.groupEnd();
      process.exit(ExitCode.error);
      return;
    }

    const defaultRole = (await pathExists(Filepathes.namespaceDefaultRolePath))
      ? await readFile(Filepathes.namespaceDefaultRolePath, 'utf8')
      : defaultRoleYml;

    const questions = [
      {
        type: 'confirm',
        name: 'createServiceAccount',
        message: 'Create a service account?',
        default: true,
      },
      {
        type: 'input',
        name: 'serviceAccountName',
        message: `The name of the service account?`,
        default: `${args.name}-deploy`,
        when: (answers: PromptAnswers) => answers.createServiceAccount,
        validate: (input: string) => !!input || 'Please enter a name.',
      },
      {
        type: 'editor',
        name: 'role',
        message: 'Edit the role for the service account.',
        default: defaultRole,
        when: (answers: PromptAnswers) => answers.createServiceAccount,
      },
      {
        type: 'confirm',
        name: 'saveRole',
        message: 'Save the given role as new default?',
        default: false,
        when: (answers: PromptAnswers) => answers.createServiceAccount,
      },
    ];

    const answers = (args.noInteraction
      ? {
        createServiceAccount: true,
        serviceAccountName: `${args.name}-deploy`,
        role: defaultRole,
        saveRole: false,
      }
      : await prompt(questions)) as PromptAnswers;

    if (answers.createServiceAccount && answers.role && answers.saveRole) {
      await ensureFile(Filepathes.namespaceDefaultRolePath);
      await writeFile(Filepathes.namespaceDefaultRolePath, answers.role);
      console.log(
        'The default role was saved under: ~/.kube/k8s-helpers/namespace/default-role.yml',
      );
      console.log('If you want to reset it, just delete the file.');
    }

    if (
      !args.noInteraction &&
      !(await simpleConfirm(
        `Create namespace "${chalk.yellow(
          args.name,
        )}" on context "${chalk.yellow(context)}"${
          answers.createServiceAccount
            ? `, with service account "${chalk.yellow(
                answers.serviceAccountName,
              )}"`
            : ''
        }. Proceed?`,
        true,
      ))
    ) {
      console.log('Aborting');
      console.groupEnd();
      return;
    }

    const code = await spawn(
      'kubectl',
      RcFile.getKubectlCtxArguments(args, ['create', 'ns', args.name]),
    );
    if (code !== 0) {
      console.error(
        chalk.red(
          'An error happend during the kubectl create namespace command.',
        ),
      );
      console.groupEnd();
      process.exit(ExitCode.error);
      return;
    }

    console.log(chalk.green(`Namespace "${args.name}" created.`));
    if (!answers.createServiceAccount) {
      console.groupEnd();
      return;
    }

    const role = datasubst(answers.role, { NAME: args.name });
    const roleNameRegex = /^\s*name:\s*(.*)$/gm.exec(role);
    if (!roleNameRegex || !roleNameRegex[1]) {
      console.error(chalk.red('No valid role name provided. aborting.'));
      console.groupEnd();
      process.exit(ExitCode.error);
      return;
    }
    const roleName = roleNameRegex[1];

    const templates = [
      role,
      serviceAccount(args.name, answers.serviceAccountName),
      rolebinding(args.name, answers.serviceAccountName, roleName),
    ];

    console.group(
      chalk.blue(`Executing <echo "templates" | kubectl create -f ->`),
    );

    try {
      const result = await exec(
        `echo "${templates.join(`${EOL}---${EOL}`)}" | kubectl create -f -`,
      );
      console.log(result);
    } catch (e) {
      console.error(chalk.red(`Error: ${e}`));
    }
    console.groupEnd();

    console.log(
      chalk.green(`ServiceAccount "${answers.serviceAccountName}" created.`),
    );
    console.log(chalk.green(`Role and RoleBinding "${roleName}" created.`));

    await namespaceKubeConfigCommand.handler({
      ...args,
      namespace: args.name,
      serviceAccount: answers.serviceAccountName,
    });

    console.groupEnd();
  },
};
