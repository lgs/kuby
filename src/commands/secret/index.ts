import { Argv, CommandModule, showHelp } from 'yargs';

import { secretBasicAuthCommand } from './basic-auth';
import { secretCreateCommand } from './create';
import { secretDockerRegistryCommand } from './docker-registry';

const secretCommands: CommandModule<any, any>[] = [secretBasicAuthCommand, secretCreateCommand, secretDockerRegistryCommand];

export const secretCommand: CommandModule = {
  command: 'secret',
  describe: 'Utilities for kubernetes secrets.',

  builder: (argv: Argv) => secretCommands.reduce((_, cur) => argv.command(cur), argv),

  handler(): void {
    showHelp('log');
  },
};
