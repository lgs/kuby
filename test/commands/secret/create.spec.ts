import { secretCreateCommand as cmd } from '../../../src/commands/secret/create';
import { KubernetesApi } from '../../../src/utils/kubernetes-api';
import { Logger } from '../../../src/utils/logger';
import { clearGlobalMocks } from '../../helpers';

describe('commands / secret / create', () => {
  beforeAll(() => {
    process.exit = jest.fn() as any;
  });

  beforeEach(() => {
    (KubernetesApi as any).create();
  });

  afterEach(() => {
    clearGlobalMocks();
  });

  it('should fail when no data is given', async () => {
    await cmd.handler({ name: 'name', data: [] } as any);
    expect(process.exit).toHaveBeenCalledWith(1);
    expect((Logger as any).instance.error).toHaveBeenLastCalledWith(
      'No data given. Must be in format key=value or key="value value". Aborting.',
    );
  });

  it('should create a namespaced secret', async () => {
    await cmd.handler({ name: 'name', data: [{ name: 'foo', value: 'bar' }] } as any);
    expect(((KubernetesApi as any).instance.core.createNamespacedSecret as jest.Mock).mock.calls[0]).toMatchSnapshot();
    expect(process.exit).not.toHaveBeenCalled();
  });

  it('should log the error on a failure', async () => {
    ((KubernetesApi as any).instance.core.createNamespacedSecret as jest.Mock).mockRejectedValueOnce({
      body: { message: 'error' },
    });
    await cmd.handler({ name: 'name', data: [{ name: 'foo', value: 'bar' }] } as any);
    expect(process.exit).toHaveBeenLastCalledWith(1);
    expect((Logger as any).instance.error).toHaveBeenLastCalledWith('error');
  });
});
