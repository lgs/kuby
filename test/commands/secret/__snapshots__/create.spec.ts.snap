// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`commands / secret / create should create a namespaced secret 1`] = `
Array [
  "default",
  Object {
    "apiVersion": "v1",
    "kind": "Secret",
    "metadata": Object {
      "name": "name",
    },
    "stringData": Object {
      "foo": "bar",
    },
    "type": "Opaque",
  },
]
`;
